pgjdbc/src/main/java/org/postgresql/jdbc/PgSQLInput.java

```diff
@@ -0,0 +1,328 @@
+  }
+
+  @Override
+  public @Nullable byte[] readBytes() throws SQLException {
+    return getNextValue(bytesConv);
+  }
+
+  @Override
+  public @Nullable Date readDate() throws SQLException {
+    return getNextValue(dateConv);
+  }
+
+  @Override
+  public @Nullable Time readTime() throws SQLException {
+    return getNextValue(timeConv);
+  }
+
+  @Override
+  public @Nullable Timestamp readTimestamp() throws SQLException {
+    return getNextValue(timestampConv);
+  }
+
+  @Override
+  public Reader readCharacterStream() throws SQLException {
+    return new StringReader(readString());
+  }
+
+  @Override
+  public InputStream readAsciiStream() throws SQLException {
+    String result = readString();
+    return new ByteArrayInputStream(result == null ? new byte[0] : result.getBytes(US_ASCII));
+  }
+
+  @Override
+  public InputStream readBinaryStream() throws SQLException {
+    return new ByteArrayInputStream(readBytes());
+  }
+
+  @Override
+  public @Nullable Object readObject() throws SQLException {
+    return getNextValue(stringConv);
+  }
+
+  @Override
+  public @Nullable <T> T readObject(Class<T> type) throws SQLException {
+    return getNextValue(getConverter(type));
+  }
+
+  @Override
+  public Ref readRef() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "readRef()");
+  }
+
+  @Override
+  public Blob readBlob() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "readBlob()");
+  }
+
+  @Override
+  public Clob readClob() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "readClob()");
+  }
+
+  @Override
+  public Array readArray() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "readArray()");
+  }
+
+  @Override
+  public @Nullable URL readURL() throws SQLException {
+    return getNextValue(urlConv);
+  }
+
+  @Override
+  public SQLXML readSQLXML() throws SQLException {
+    return new PgSQLXML(connection, readString());
+  }
+
+  @Override
+  public RowId readRowId() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "getRowId()");
+  }
+
+  @Override
+  public boolean wasNull() throws SQLException {
+    return wasNull == null ? false : wasNull;
+  }
+
+  @Override
+  public NClob readNClob() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "readNClob()");
+  }
+
+  @Override
+  public String readNString() throws SQLException {
+    throw Driver.notImplemented(this.getClass(), "readNString()");
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> SQLFunction<String, T> getConverter(Class<T> type) throws SQLException {
+    if (type.isArray()) {
+      Class<?> itemType = type.getComponentType();
+      SQLFunction<String, ?> converter = getConverter(itemType);
+      return (value) -> {
+        List<String> items = new SQLDataReader().parseArray(value);
+        Object results = java.lang.reflect.Array.newInstance(itemType, items.size());
```

Comment:
Do we know how well this plays with Graalvm native ?

Reply:
I don't.